src/abstract-interpretation/domain.ts

```diff
@@ -0,0 +1,166 @@
+	Min,
+	/** The bound that's inclusive is the greater one */
+	Max,
+	/** Equality is only based on the "raw" values */
+	IgnoreInclusivity
+}
+
+function compareIntervals(compareType: CompareType, interval1: IntervalBound, interval2: IntervalBound): number {
+	const diff = interval1.value - interval2.value
+	if(diff !== 0 || compareType === CompareType.IgnoreInclusivity) {
+		return diff
+	}
+	switch(compareType) {
+		case CompareType.Min:
+			return Number(!interval1.inclusive) - Number(!interval2.inclusive)
+		case CompareType.Max:
+			return Number(interval1.inclusive) - Number(interval2.inclusive)
+		default:
+			assertUnreachable(compareType)
+	}
+}
+
+function compareIntervalsByTheirMinimum(interval1: Interval, interval2: Interval): number {
+	return compareIntervals(CompareType.Min, interval1.min, interval2.min)
+}
+
+function compareIntervalsByTheirMaximum(interval1: Interval, interval2: Interval): number {
+	return compareIntervals(CompareType.Max, interval1.max, interval2.max)
+}
+
+export function doIntervalsOverlap(interval1: Interval, interval2: Interval): boolean {
+	const diff1 = compareIntervals(CompareType.IgnoreInclusivity, interval1.max, interval2.min)
+	const diff2 = compareIntervals(CompareType.IgnoreInclusivity, interval2.max, interval1.min)
+
+	// If one interval ends before the other starts, they don't overlap
+	if(diff1 < 0 || diff2 < 0) {
+		return false
+	}
+	// If their end and start are equal, they only overlap if both are inclusive
+	if(diff1 === 0) {
+		return interval1.max.inclusive && interval2.min.inclusive
+	}
+	if(diff2 === 0) {
+		return interval2.max.inclusive && interval1.min.inclusive
+	}
+
+	return true
+}
+
+export function unifyDomains(domains: Domain[]): Domain {
+	const unifiedIntervals = unifyIntervals(domains.flatMap(domain => Array.from(domain.intervals)))
+	return new Domain(...unifiedIntervals)
+}
+
+export function unifyIntervals(intervals: Interval[]): Interval[] {
+	const sortedIntervals = intervals.sort(compareIntervalsByTheirMinimum)
+	if(sortedIntervals.length === 0) {
+		return []
+	}
+
+	const unifiedIntervals: Interval[] = []
+	let currentInterval = sortedIntervals[0]
+	for(const nextInterval of sortedIntervals) {
+		if(doIntervalsOverlap(currentInterval, nextInterval)) {
+			const intervalWithEarlierStart = compareIntervalsByTheirMinimum(currentInterval, nextInterval) < 0 ? currentInterval : nextInterval
+			const intervalWithLaterEnd = compareIntervalsByTheirMaximum(currentInterval, nextInterval) > 0 ? currentInterval : nextInterval
+			currentInterval = new Interval(intervalWithEarlierStart.min, intervalWithLaterEnd.max)
+		} else {
+			unifiedIntervals.push(currentInterval)
+			currentInterval = nextInterval
+		}
+	}
+	unifiedIntervals.push(currentInterval)
+	return unifiedIntervals
+}
+
+export function addDomains(domain1: Domain, domain2: Domain): Domain {
+	const intervals = new Set<Interval>()
+	for(const interval1 of domain1.intervals) {
```

Comment:
Maybe we should abstract away the cartesian product to only define max and min.

Reply:
Tracked in #608